File: Sheets/src/neetcode/practice/Backtracking/_2_combinationSum.java
package neetcode.practice.Backtracking;

import java.util.ArrayList;
import java.util.List;

class Solution {
  final public List<List<Integer>> result = new ArrayList<>();
  final public List<Integer> augment = new ArrayList<>();

  public List<List<Integer>> combinationSum(int[] nums, int target) {
    this.dfs(0, 0, target, nums);
    return this.result;
  }

  private void dfs(int index, int currentSum, int target, int[] nums) {
    if (currentSum == target) {
      this.result.add(new ArrayList<>(this.augment));
      return;
    }

    if ((index >= nums.length) || currentSum > target)
      return;

    this.augment.add(nums[index]);
    this.dfs(index, currentSum + nums[index], target, nums);

    this.augment.remove(this.augment.size() - 1);
    this.dfs(index + 1, currentSum, target, nums);
  }
}

File: Sheets/src/neetcode/practice/Backtracking/_2_CombinationSum.cc
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>> res{};
  int target;

  void search(int sum, const std::vector<int> &nums, std::vector<int> cur,
              size_t idx...) {
    if (sum == target) {
      this->res.push_back(cur);
      return;
    }

    if (sum > target || idx == nums.size())
      return;

    cur.push_back(nums[idx]);
    search(sum + nums[idx], nums, cur, idx);
    cur.pop_back();
    search(sum, nums, cur, idx + 1);
  }

  std::vector<std::vector<int>> combinationSum(std::vector<int> &nums,
                                               int target) {
    this->target = target;
    search(0, nums, {}, 0);
    return this->res;
  }
};
